### ISSUE TITLE
Minimal repro app for a Direct3D 12 driver crash (debugger) and device removal (no debugger) during PSO creation

### ISSUE THREAD

**Issue by mightycow** [IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT#817]
### Checklist [README]

- [X] Device is not undervolted nor overclocked
- [X] Device is using the [latest drivers](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/wiki/Intel-driver-updates-and-changelogs)
- [X] Application is not cracked, modded and use the latest patch

### Application [Required]

None, this is a minimal repro app for the driver team: https://myt.playmorepromode.com/intel/

### Processor / Processor Number [Required]

Core i5-13600K

### Graphic Card [Required]

UHD Graphics 770

### GPU Driver Version [Required]

31.0.101.5594

### Other GPU Driver version

The problem happens with 5594, 5590 and 5444

### Rendering API [Required]

- [ ] Vulkan
- [ ] OpenGL
- [X] DirectX12
- [ ] DirectX11
- [ ] DirectX10
- [ ] DirectX9
- [ ] Not applicable

### Windows Build Number [Required]

- [ ] Windows 11 23H2
- [ ] Windows 11 22H2
- [ ] Windows 11 21H2
- [X] Windows 10 22H2
- [ ] Windows 10 21H2
- [ ] Other (Please specify)

### Intel System Support Utility report

[igcit_ssu.txt](https://github.com/user-attachments/files/16147896/igcit_ssu.txt)

### Description and steps to reproduce [Required]

1. Download source code from https://myt.playmorepromode.com/intel/ or from the "Crash dumps" section below
2. Run through VC++ debugger to trigger a crash in igc1464.dll
3. Run without a debugger to trigger a device removal (0x887A0005 from CreateComputePipelineState)

### Crash dumps [Required, if applicable]

Here's the minimal repro app with full source code and VC++ 2022 project files.
[intel_driver_crash_d3d12_no_exe_no_dll.zip](https://github.com/user-attachments/files/16147907/intel_driver_crash_d3d12_no_exe_no_dll.zip)
The app makes sure to only create a device on an Intel adapter.
All it does then is create a PSO using the supplied DXIL byte code.
Agility SDK headers are included but the DLLs are missing. The version used is 608.
Archive versions with binaries are hosted here: https://myt.playmorepromode.com/intel/

**Comment by Karen-Intel** [comment]
Hey @mightycow thank you for your submission. I'll be assisting you on these
We just got a new driver yesterday and one of the newest machines. I'll do some testing and get back to you for confirmation/questions. Stay tuned!

Karen

**Comment by Karen-Intel** [comment]
Hii @mightycow update!
I was able to reproduce the crash in the igc1464.dll, but I'm trying to run the same project on a similar system with another GPU (for debug purposes and issue isolation)
I'm getting a syntax error: expecting a type specification near "UINT" when I run it in my 3050. Is this expected? 
![NVIDIA 3050 Driver Crash Project Test](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/assets/91162218/fec679ef-53ff-4a6c-9319-ad4d7c612682)

Thanks for your help

Karen

**Comment by mightycow** [comment]
Hello @Karen-Intel,

Thanks for taking a look at my report.

Here are my thoughts and some extra information:
- MDL2025 is a build error, not a run-time error. I can't tell exactly what's going on because the full build error is not shown (which line is the problem, what header is included from where). It might be due to an incompatible mix between local headers (Agility SDK file in the project's folder) and system headers (Platform SDK). I can't say that build error was expected: I only have the one machine. I do have a GPU from each IHV, however.
- To run on a non-Intel GPU, you need to change the source code. The code I supplied only creates a device on an Intel adapter.
- This problem also happens on AMD (24.5.1) but doesn't on NVIDIA (551.76).
- The problem on AMD was confirmed to me to be a shader compiler issue by an AMD employee.
- Speculation: I didn't ask for details but it's quite probable that the problem is in LLVM code that both AMD and Intel use in their shader compiler infrastructure.
- Speculation: The debugger/no-debugger behavior difference can also be explained by the presence of an exception filter installed by the Intel driver that allows the debugger a chance to run its own handler. A device removal is more graceful than a crash but it's still not particularly user-friendly.

**Comment by mightycow** [comment]
Instead of spending time fixing the build issue on the second machine, you can copy over the compiled binaries from the first machine.

App output when it fails to create the PSO:
```
Error code 0x887A0005 (The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action.)
device->CreateComputePipelineState(&desc, IID_PPV_ARGS(&pso)) @ App::CreatePSO:211
Press any key to continue . . .
```

App output when it succeeds:
```
Compiled without a crash! Hurray!
```

There are 2 lines to change to select another vendor's adapter: 104 and 119.
104: `if(SUCCEEDED(adapter->GetDesc1(&desc)) && desc.VendorId == VENDORID_INTEL)`
119: `assert(desc.VendorId == VENDORID_INTEL);`

Hope that helps.

**Comment by Karen-Intel** [comment]
Hey @mightycow thank you for the details!

After exporting the executable from my Intel machine (where it fails) and after changing the VendorID, I was able to get the: Compiled without a crash! Hurray! in my NVIDIA 3050

I will proceed to create an internal report. Will be back tomorrow with the number for your reference.

Karen

**Comment by mightycow** [comment]
Thank you @Karen-Intel

**Comment by Karen-Intel** [comment]
Hii @mightycow!

Sharing the internal ID of the report I have submitted: 14022859473

Driver general cadence for fixes takes between 3-6 months avg, however some fixes take less than that. 
Any update I'll post it in this thread

Thank you so much for your help!

Karen

**Comment by mightycow** [comment]
Thanks, @Karen-Intel

**Comment by SavageWaffle** [comment]
Has this been resolved?

**Comment by Felipe-Intel** [comment]
@SavageWaffle this is still under investigation

**Comment by Karen-Intel** [comment]
@mightycow update on this case: the fix for this issue has been released :)
Could you help us verify the issue is not happening anymore with driver [v.32.0.101.7026](https://www.intel.com/content/www/us/en/download/785597/intel-arc-iris-xe-graphics-windows.html) and let me know?

Thanks a lot 

Karen

**Comment by mightycow** [comment]
Hello @Karen-Intel ,

Thanks for letting me know.
So, I have good news and bad news:
- Old version 31.0.101.5594 (0x001F0000006515DA)
  - Confirmed crash in the minimal repro app
  - Confirmed crash in my rendering engine
  - Crash location and type: igc1464.dll, Access violation reading location
- New version 32.0.101.7026 (0x0020000000651B72)
  - Confirmed no crash in the minimal repro app
  - Confirmed no crash in the rendering engine when creating the PSO
  - Cannot confirm the compiled PSO has valid behavior
  - Rendering even the simplest of PSOs in the affected rendering pipeline will now trigger TDR:
    - DXGI_ERROR_DEVICE_HUNG / DEVICE_REMOVAL_PROCESS_AT_FAULT
    - DXGI_ERROR_DRIVER_INTERNAL_ERROR / DEVICE_REMOVAL_PROCESS_POSSIBLY_AT_FAULT

Unfortunately, the current state of affairs on my end is a big regression but I don't know who's at fault yet.
I will try to figure out why the crash happens with Microsoft's DRED since Intel has no equivalent to NVIDIA Aftermath or AMD FidelityFX Breadcrumbs that I know of. If Intel has one, please let me know.
This might have to wait a few days as I have other things cooking in the oven...

tl;dr
- Confirmed no crash during PSO creation
- Did NOT confirm the PSO is valid (expected behavior & no crash during execution)

**Comment by Karen-Intel** [comment]
@mightycow is OK. Remember our focus is the crash of the igc1464.dll :) Thanks

**Comment by mightycow** [comment]
Hello @Karen-Intel,

> Remember our focus is the crash of the igc1464.dll

I _very_ strongly disagree. The focus is making sure `ID3D12Device::CreateComputePipelineState` with the data given above (byte code, pipeline state, root signature, etc.) behaves _correctly_. Not crashing is an extremely low bar that the new driver has cleared but I don't have confirmation that the behavior is _correct_.

If the following 2 statements are true, you can go ahead and close the issue:

1. The supplied DXIL byte code is valid.
    If not, then the new driver is still not behaving correctly because it fails to return an error code.
2. The generated GPU instruction stream is correct.
    I cannot confirm it yet and there has been no assurances or proof from the driver team.
    Assurances are nice but you know the old saying: "trust, but verify".

From my personal point of view, this issue is not over until the moment I see that compute shader executing correctly on an Intel device.

**Comment by Karen-Intel** [comment]
@mightycow There's no rush to close this one out, you can relax. 
We can wait until you complete your other tasks and verify if the compiled PSO has valid behavior :)

Keep us posted

**Comment by Karen-Intel** [comment]
@mightycow do you have an update about your testing? Our team is asking for status
If there's no reply soon we'll close this thread 

Thanks

Karen

**Comment by mightycow** [comment]
Hello @Karen-Intel,

I've not been able to spend that much time on it but made very good progress on both issues.
Issue 1 being the main one from this thread and issue 2 being the new crash I get from the latest WHQL drivers.

Issue 1
- I've found a way to sidestep issue 2 so I can test the shader in question in context.
- Unfortunately, my Intel UHD 770 is not fast enough to handle any valid dataset I have so the TDR mechanism is triggered.
- I can hopefully generate a really low-fidelity new one and get things going.

Issue 2
- I know the new crash is a page fault (thanks to DRED) and which draw call is the last successful one.
- It only happens when certain CPU writes happen to a very specific CPU-visible position vertex buffer.
- None of the CPU writes are out of bound - confirmed with both new CPU code and a RenderDoc capture.
- However, I think I have just caught the fact that a single CPU write that is within a range of 1-2 draw calls of the DRED threshold has an offset that goes back to 0, i.e. it seems that the writes are not contiguous as they ideally should be.
(I would expect this to be bad for perf. but certainly not to crash.)

I know issue 2 is not your concern in this post but it has prevented me from checking the fix of issue 1 so I couldn't really ignore it.

In any case, I think I can get a test with new data that won't TDR tomorrow.

tl;dr:
- should have a definitive answer within 24 hours
- will file a report for the new crash (issue 2) later when I have everything confirmed and a minimal repro app ready

**Comment by mightycow** [comment]
Hello @Karen-Intel,

I have finally managed to validate the fix on my Intel UHD 770.
